File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/MessageIterator.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.MessageIterator.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu;

import java.util.List;
import java.util.Iterator;

public class MessageIterator implements Iterator<Message> {

	private Iterator<Message> _iter;
	
	public MessageIterator(List<Message> list)
	{
		this._iter = list.iterator();
	}
	
	public boolean hasNext()
	{
		return this._iter.hasNext();
	}
	
	public Message next()
	{
		return this._iter.next();
	}
	
	public void remove()
	{
		this._iter.remove();
	}
}

File: BEmu_cpp/src/BloombergTypes/MessageIterator.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/BloombergTypes/MessageIterator.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include "BloombergTypes/Message.h"
#include "BloombergTypes/MessagePtr.h"
#include "BloombergTypes/MessageIterator.h"
#include "BloombergTypes/Event.h"
#include "BloombergTypes/EventPtr.h"
#include "BloombergTypes/ElementPtr.h"
#include <stack>

namespace BEmu
{
	MessageIterator::MessageIterator(const Event& evt)
	{
		this->_position = -1;

		boost::shared_ptr<EventPtr> evtP(evt.getEventPtr());

		std::vector< boost::shared_ptr<MessagePtr> > list = evtP->getMessages();
		
		for(std::vector< boost::shared_ptr<MessagePtr> >::const_iterator iter = list.begin(); iter != list.end(); ++iter)
		{
			boost::shared_ptr<MessagePtr> msg = *iter;

			this->_listP.push_back(msg);
		}
	}

	MessageIterator::~MessageIterator()
	{
		this->_listP.clear();
	}

	bool MessageIterator::next()
	{
		this->_position++;
		return this->_position < (int)(this->_listP.size());
	}

	bool MessageIterator::isValid() const
	{
		return this->_position >= 0 && this->_position < (int)(this->_listP.size());
	}

	Message MessageIterator::message(bool createClonable) const
	{
		boost::shared_ptr<MessagePtr> messageP = this->_listP.at(this->_position);
		Message msg(messageP);
		return msg;
	}
}